pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java

```diff
@@ -1046,6 +1064,261 @@ public boolean needReloadSegments()
     return needReload;
   }
 
+  @Override
+  public List<NeedRefreshResponse> getSegmentsForRefresh(TableConfig tableConfig, Schema schema) {
+    List<NeedRefreshResponse> segmentsRequiringRefresh = new ArrayList<>();
+    List<SegmentDataManager> segmentDataManagers = acquireAllSegments();
```

Comment:
acquireAllSegments also increases the ref count
so we need to decrease it as well once we get the list

Reply:
Also, since we are running a for loop, wouldn't a better way be to acquire each segment one by one

Reply:
I am releasing the segments now. To acquire a segment, how do I get the segment name ? One option is to acquire all segments, convert that into a list of names and then reacquire one by one. Any other options ?

Reply:
Yes. Also new segments will be missed. That is a reasonable compromise ?

Reply:
@vrajat and I chatted . I think we can start with current code (acquireAllSegments) . These checks are local and in-memory and should be quick. We can record the start/end times for this operation. If needed we can optimize this further.

Reply:
Cool, I am fine with that!